tools/validate/validate.cpp

```diff
@@ -0,0 +1,197 @@
+                        static_cast<unsigned int>(ch));
+                }
+            }
+
+            if ((prev == ' ' || prev == '\t') && (ch == CR || ch == LF)) {
+                ++trailing_whitespace_lines;
+            }
+
+            (void) exchange(previous3, exchange(previous2, exchange(prev, ch)));
+        }
+    }
+
+    if (prev == CR) { // file ends with CR
+        has_cr = true;
+    }
+
+    if (has_cr) {
+        fprintf(stderr, "Validation failed: %s contains CR line endings (possibly damaged CRLF).\n", filepath.c_str());
+    } else if (has_lf && has_crlf) {
+        fprintf(stderr, "Validation failed: %s contains mixed line endings (both LF and CRLF).\n", filepath.c_str());
+    } else if (has_lf) {
+        fprintf(stderr, "Validation failed: %s contains LF line endings.", filepath.c_str());
+
+        if (prev != LF) {
+            fprintf(stderr, " Also, it doesn't end with a newline.\n");
+        } else if (previous2 == LF) {
+            fprintf(stderr, " Also, it ends with multiple newlines.\n");
+        } else {
+            fprintf(stderr, "\n");
+        }
+    } else if (has_crlf) {
+        if (previous2 != CR || prev != LF) {
+            fprintf(stderr, "Validation failed: %s doesn't end with a newline.\n", filepath.c_str());
+        } else if (previous3 == LF) {
+            fprintf(stderr, "Validation failed: %s ends with multiple newlines.\n", filepath.c_str());
+        }
+    } else {
+        fprintf(stderr, "Validation failed: %s doesn't contain any newlines.\n", filepath.c_str());
+    }
+
+    if (has_utf8_bom) {
+        fprintf(stderr, "Validation failed: %s contains UTF-8 BOM characters.\n", filepath.c_str());
+    }
+
+    if (tab_characters != 0) {
+        fprintf(stderr, "Validation failed: %s contains %zu tab characters.\n", filepath.c_str(), tab_characters);
+    }
+
+    if (trailing_whitespace_lines != 0) {
+        fprintf(stderr, "Validation failed: %s contains %zu lines with trailing whitespace.\n", filepath.c_str(),
+            trailing_whitespace_lines);
+    }
+}
+
+int main() {
+    static constexpr array skipped_directories{
+        ".git"sv,
+        ".vs"sv,
+        ".vscode"sv,
+        "out"sv,
+        "vcpkg"sv,
+    };
+
+    static constexpr array skipped_extensions{
+        ".dll"sv,
+        ".exe"sv,
+        ".obj"sv,
+    };
+
+    // TRANSITION, P0202R3, use constexpr is_sorted()
+    assert(is_sorted(skipped_directories.begin(), skipped_directories.end()));
+    assert(is_sorted(skipped_extensions.begin(), skipped_extensions.end()));
+
+    vector<unsigned char> buffer; // reused for performance
+
+    try {
+        for (filesystem::recursive_directory_iterator rdi{"."}, last; rdi != last; ++rdi) {
+            if (!rdi->is_regular_file()) {
+                if (rdi->is_directory()) {
+                    const string filename = rdi->path().filename().string();

```

Comment:
I think you should either `u8string` or `native`.

Reply:
does it matter? we don't call the wide version of fopen anyway, so nothing will work with unicode filenames unless you turned on the utf-8 windows setting.

Reply:
>we don't call the wide version of fopen anyway

Maybe we should fix that too. I wouldn't want this to fail if we have a customer in Japan or China where string() is going to want Shift-JIS or a GBnnnn standard.

Reply:
well the narrow one will work for those won't it?

Reply:
No. For example `\` is `¥` in the Japanese locales.

Reply:
I'd actually like to validate that filenames in our tree are printable ASCII, but the directory structure above can be arbitrary Unicode.

Reply:
In that case I still think you should use u8path, and add post validation to that if you want.

Reply:
I've changed this to pass the native `wchar_t` string to `_wfopen()`, and to print it in diagnostics with `fwprintf()`. I am not yet attempting to activate the UTF-16 codepath for `fwprintf()`, so paths with actually-wide characters will be correctly opened, but may not be correctly printed in diagnostics. Please meow if you want further changes.